x-pack/plugins/security/server/audit/audit_events.ts

```diff
@@ -0,0 +1,245 @@
+    /**
+     * Saved object that was created, changed, deleted or accessed as part of the action.
+     */
+    saved_object?: {
+      type: string;
+      id?: string;
+    };
+    /**
+     * Any additional event specific fields.
+     */
+    [x: string]: any;
+  };
+  error?: {
+    code?: string;
+    message?: string;
+  };
+  http?: {
+    request?: {
+      method?: string;
+      body?: {
+        content: string;
+      };
+    };
+    response?: {
+      status_code?: number;
+    };
+  };
+  url?: {
+    domain?: string;
+    full?: string;
+    path?: string;
+    port?: number;
+    query?: string;
+    scheme?: string;
+  };
+}
+
+export enum EventCategory {
+  DATABASE = 'database',
+  WEB = 'web',
+  IAM = 'iam',
+  AUTHENTICATION = 'authentication',
+  PROCESS = 'process',
+}
+
+export enum EventType {
+  USER = 'user',
+  GROUP = 'group',
+  CREATION = 'creation',
+  ACCESS = 'access',
+  CHANGE = 'change',
+  DELETION = 'deletion',
+}
+
+export enum EventOutcome {
+  SUCCESS = 'success',
+  FAILURE = 'failure',
+  UNKNOWN = 'unknown',
+}
+
+export interface HttpRequestParams {
+  request: KibanaRequest;
+}
+
+export function httpRequestEvent({ request }: HttpRequestParams): AuditEvent {
+  const [path, query] = request.url.path?.split('?') ?? [];
+
+  return {
+    message: `User is requesting [${path}] endpoint`,
+    event: {
+      action: 'http_request',
+      category: EventCategory.WEB,
+      outcome: EventOutcome.UNKNOWN,
+    },
+    http: {
+      request: {
+        method: request.route.method,
+      },
+    },
+    url: {
+      domain: request.url.hostname,
+      path,
+      port: request.url.port ? parseInt(request.url.port, 10) : undefined,
+      query,
+      scheme: request.url.protocol,
+    },
+  };
+}
+
+export interface UserLoginParams {
+  authenticationResult: AuthenticationResult;
+  authenticationProvider?: string;
+  authenticationType?: string;
+}
+
+export function userLoginEvent({
+  authenticationResult,
+  authenticationProvider,
+  authenticationType,
+}: UserLoginParams): AuditEvent {
+  return {
+    message: authenticationResult.user
+      ? `User [${authenticationResult.user.username}] has logged in using ${authenticationType} provider [name=${authenticationProvider}]`
+      : `Failed attempt to login using ${authenticationType} provider [name=${authenticationProvider}]`,
+    event: {
+      action: 'user_login',
+      category: EventCategory.AUTHENTICATION,
+      outcome: authenticationResult.user ? EventOutcome.SUCCESS : EventOutcome.FAILURE,
+    },
+    user: authenticationResult.user && {
+      name: authenticationResult.user.username,
+      roles: authenticationResult.user.roles,
+    },
+    kibana: {
+      authentication_provider: authenticationProvider,
+      authentication_type: authenticationType,
+      authentication_realm: authenticationResult.user?.authentication_realm.name,
+      lookup_realm: authenticationResult.user?.lookup_realm.name,
```

Comment:
My understanding of the field was that you always have a current space in your session (defaulting to `default` space). So this field doesn't so much say that you logged into that space but that that's the space currently assigned to the session. Is that incorrect? 
Happy to remove it but it's automatically populated when available by the audit service so I'd need to explicitly set this to something empty for authentication events which feels unnecessary as it doesn't really hurt having it populated.

Reply:
> My understanding of the field was that you always have a current space in your session (defaulting to default space). So this field doesn't so much say that you logged into that space but that that's the space currently assigned to the session. Is that incorrect?

_Typically_ that's true, but authentication is a bit of a special case. A successfully authenticated user may not be authorized to access the `default` space, so having their authentication event mention this space is potentially misleading.

From the server's perspective, the user appears to be in the default space, since the login URL (`/login`) is not scoped to a particular space. The absence of a space identifier in the URL generally means you're in the `default` space, but I don't consider our authentication flow to be part of any one space.

We unfortunately don't have a mechanism today to distinguish between the default space, and "outside" of any space.

Reply:
Got you, that makes sense to me. I'll leave it blank for authentication events in that case.